### PR TITLE
Add SDP3x sensor

### PR DESCRIPTION
**Pull request by Azimath**
# What does this implement/fix? 

Adds a component to support the SDP3x sensor, an I2C differential pressure sensor useful for measuring small pressure differences like the drop across an HVAC filter.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1333

## Test Environment

- [x] ESP32
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
  - platform: sdp3x
    name: "HVAC Filter Pressure drop"
    id: filter_pressure
    update_interval: 5s
    accuracy_decimals: 3
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/sdp3x/sdp3x.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+static const char *const TAG = "sdp3x.sensor";
+static const uint8_t SDP3X_SOFT_RESET[2] = {0x00, 0x06};
+static const uint8_t SDP3X_READ_ID1[2] = {0x36, 0x7C};
+static const uint8_t SDP3X_READ_ID2[2] = {0xE1, 0x02};
+static const uint8_t SDP3X_START_DP_AVG[2] = {0x36, 0x15};
+static const uint8_t SDP3X_STOP_MEAS[2] = {0x3F, 0xF9};
+
+void SDP3XComponent::update() {
+  this->set_timeout("pressure", 5, [this]() { this->read_pressure_(); });
```
Comment (jesserockz) [inline esphome/components/sdp3x/sdp3x.cpp]: I don't really see the need for this?

```suggestion
  this->read_pressure_();
```

File: esphome/components/sdp3x/sensor.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import esphome.config_validation as cv
+from esphome.components import i2c, sensor
+from esphome.const import (
+    CONF_ID,
+    DEVICE_CLASS_PRESSURE,
+    ICON_EMPTY,
+    UNIT_HECTOPASCAL,
+)
+
+DEPENDENCIES = ["i2c"]
```
Comment (jesserockz) [inline esphome/components/sdp3x/sensor.py:12]: Please add yourself as a codeowner and run `script/build_codeowners.py`
```suggestion
CODEOWNERS = ["@Azimath"]
DEPENDENCIES = ["i2c"]
```

File: esphome/components/sdp3x/sensor.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    .extend(cv.polling_component_schema("60s"))
+    .extend(i2c.i2c_device_schema(0x21))
+)
+
+
+def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    yield cg.register_component(var, config)
+    yield i2c.register_i2c_device(var, config)
+    yield sensor.register_sensor(var, config)
```
Comment (jesserockz) [inline esphome/components/sdp3x/sensor.py]: **Optional** change but a good place to inform you of the async changes in the codebase =)
```suggestion
async def to_code(config):
    var = cg.new_Pvariable(config[CONF_ID])
    await cg.register_component(var, config)
    await i2c.register_i2c_device(var, config)
    await sensor.register_sensor(var, config)
```

File: esphome/components/sdp3x/sensor.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    UNIT_HECTOPASCAL,
+)
+
+DEPENDENCIES = ["i2c"]
+
+sdp3x_ns = cg.esphome_ns.namespace("sdp3x")
+SDP3XComponent = sdp3x_ns.class_("SDP3XComponent", cg.PollingComponent, i2c.I2CDevice)
+
+CONFIG_SCHEMA = (
+    sensor.sensor_schema(UNIT_HECTOPASCAL, ICON_EMPTY, 3, DEVICE_CLASS_PRESSURE)
```
Comment (jesserockz) [inline esphome/components/sdp3x/sensor.py]: Also makes this a statistic that has min/max/average in Home Assistant (needs the import above too)
```suggestion
    sensor.sensor_schema(
        UNIT_HECTOPASCAL, 
        ICON_EMPTY, 
        3, 
        DEVICE_CLASS_PRESSURE, 
        STATE_CLASS_MEASUREMENT,
    )
```

File: esphome/components/sdp3x/sensor.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
+import esphome.codegen as cg
+import esphome.config_validation as cv
+from esphome.components import i2c, sensor
+from esphome.const import (
+    CONF_ID,
+    DEVICE_CLASS_PRESSURE,
+    ICON_EMPTY,
+    UNIT_HECTOPASCAL,
```
Comment (jesserockz) [inline esphome/components/sdp3x/sensor.py:9]:
```suggestion
    STATE_CLASS_MEASUREMENT,
    UNIT_HECTOPASCAL,
```

File: esphome/components/sdp3x/sdp3x.cpp [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    return;
+  }
+
+  if (!(check_crc_(&data[0], 2, data[2]) && check_crc_(&data[3], 2, data[5]) && check_crc_(&data[6], 2, data[8]))) {
+    ESP_LOGW(TAG, "Invalid SDP3X data!");
+    this->status_set_warning();
+    return;
+  }
+
+  int16_t pressure_raw = (int16_t)((data[0] << 8) | data[1]);
```
Comment (jesserockz) [inline esphome/components/sdp3x/sdp3x.cpp]:
```suggestion
  int16_t pressure_raw = encode_uint16(data[0], data[1]);
```

File: esphome/components/sdp3x/sdp3x.cpp [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+void SDP3XComponent::dump_config() {
+  ESP_LOGCONFIG(TAG, "SDP3X:");
+  LOG_I2C_DEVICE(this);
+  if (this->is_failed()) {
+    ESP_LOGE(TAG, "Connection with SDP3X failed!");
+  }
+  LOG_UPDATE_INTERVAL(this);
+
+  LOG_SENSOR("  ", "Pressure", this);
+}
```
Comment (jesserockz) [inline esphome/components/sdp3x/sdp3x.cpp]:
```suggestion
void SDP3XComponent::dump_config() {
  LOG_SENSOR("  ", "SDP3X", this);
  LOG_I2C_DEVICE(this);
  if (this->is_failed()) {
    ESP_LOGE(TAG, "  Connection with SDP3X failed!");
  }
  LOG_UPDATE_INTERVAL(this);
}
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/sdp3x/sensor.py (hunk 5 above)
Comment (Azimath) [reply esphome/components/sdp3x/sensor.py:9]: Done, not sure why github says otherwise

File: esphome/components/sdp3x/sensor.py (hunk 2 above)
Comment (Azimath) [reply esphome/components/sdp3x/sensor.py:12]: Done, github doesn't see it as outdated for some reason

**Review by jesserockz — APPROVED** [review]

File: esphome/components/sdp3x/sdp3x.cpp [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+  if (!(check_crc_(&data[0], 2, data[2]) && check_crc_(&data[3], 2, data[5]) && check_crc_(&data[6], 2, data[8]))) {
+    ESP_LOGW(TAG, "Invalid SDP3X data!");
+    this->status_set_warning();
+    return;
+  }
+
+  int16_t pressure_raw = encode_uint16(data[0], data[1]);
+  float pressure = pressure_raw / pressure_scale_factor_;
+  ESP_LOGD(TAG, "Got raw pressure=%d, scale factor =%.3f ", pressure_raw, pressure_scale_factor_);
```
Comment (OttoWinter) [inline esphome/components/sdp3x/sdp3x.cpp]:
```suggestion
  ESP_LOGV(TAG, "Got raw pressure=%d, scale factor =%.3f ", pressure_raw, pressure_scale_factor_);
```

Nit: should be v log level

**Review by OttoWinter — COMMENTED** [review]
LGTM :+1: Thanks!

**Pull request merged**